Added Qualys CAR connector

thangaraj-ramesh opened the pull request:
Added Qualys CAR connector.

thangaraj-ramesh committed:
Add files via upload

Qualys car connector

omkar-g commented on connectors/qualys/configurations/config.json:
```diff
@@ -0,0 +1,31 @@
+{
+    "connection": {
+        "schedule": {
```
We dont need schedule field, if we are fine with default option which is the case here

thummalar replied:
Schedule field is required. The Qualys Agent will send collected information every 4 hours. With default options ( 5m,10m,...) we will get less updates from the agent, so we are configuring schedulers with minimum interval of 1h,2h,4h,6h,... to get reasonable data to import.

omkar-g commented on connectors/qualys/configurations/lang_en.json:
```diff
@@ -0,0 +1,42 @@
+{
+    "connection":{
+        "schedule": {
```
This can be removed as well.

thummalar replied:
Schedule field is required. The Qualys Agent will send collected information every 4 hours. With default options ( 5m,10m,...) we will get less updates from the agent, so we are configuring schedulers with minimum interval of 1h,2h,4h,6h,... to get reasonable data to import.

omkar-g commented on connectors/qualys/connector/data_handler.py:
```diff
@@ -0,0 +1,287 @@
+                        self.add_edge('asset_application', asset_application)
+
+                        # application vulnerability edge creation
+                        application_vulnerability = {'_from_external_id': str(app['id']),
```
Are these vulnerability specific to this application in the asset? If it is just same vuln we have for asset then we dont need edge between application_vulnerability.

thummalar replied:
Yes, these vulnerabilities are specific to this application in the asset.

Our assumption now is that if vulnerable application is present in asset,  we can map asset to vulnerability(asset_vulnerability). 'application_vulnerability' edge is not required and can be removed.

Can we retain application node?

omkar-g commented on connectors/qualys/connector/inc_import.py:
```diff
@@ -0,0 +1,168 @@
+        asset_edges = {'ipaddress': set(), 'macaddress': set(), 'hostname': set(),
+                       'account': set(), 'application': set(), 'vulnerability': set(), 'geolocation': set()}
+
+        search_result = context().car_service.graph_search('asset', str(asset_id))
```
graph_search is deprecated function shouldn't be used.

thummalar replied:
We didn't find any alternative API for graph search, please let us know which is the alternative API to use for getting asset associated node and edge information if 'graph_search' is deprecated.

thummalar replied:
We will use API 'graph_attribute_search' for getting information on nodes and use the information to disable edges

thummalar replied:
We have updated code to use the API v3 (URL : 'https://***.ibmcloudsecurity.com/api/car/v3/query' ) to get the asset edge information from CAR DB.

Note: not able to use 'graph_attribute_search' to get the edges  since we are getting 500 error code.

thangaraj-ramesh committed:
graph_search API replaced with graphQL query search

Code updated to replace 'graph_search' API with graphQL search to get the asset edges from CAR DB. Automation and unit tests are updated.
